### PR TITLE
Set final day button for front-end ui.

### PR DESCRIPTION
**Pull request by kseniaxu0704**
Task: 
create a button on the user profile page to set up the final day;

Current functions: 
1) After clicking the set final day button, it pops up a date picker(name setUpFinalDayPopUp), which is possible to select a final day.
2) The set final day button could be changed to deleted final day button if the user profile end date has been set; 
3) If click the delete the final day button, the user profile end date could be deleted, and the end date could become NA

About code:

Related component : 
1) BasicInformationTab->  2)SetUpFinalDayButton -> 3) SetUpFinalDayPopUp

Related action: 
userManagement.js (it uses Axios to patch endDate to REST API)

General logical about SetUpFinalDayButton : 
use react hook - 
1) useState(const[isSet, setIsSet]) to record if there is a final day set and
2) useState (const[finalDayDateOpen, setFinalDayDateOpen]) to control the SetUpFinalDayPopUp open or not; 

General logical about UserManager : 
it is related to controller - UserProfileController-change user status; 
React route : path="/userprofile/:userId" function :patch(controller.changeUserStatus);

For the next step: if reach out on the final day, the user will automatically change from active status to inactive and send an email to the admin to notify.

### CODE REVIEW

File: src/actions/userManagement.js [hunk 1]
```diff
@@ -7,9 +7,10 @@ import {
   USER_PROFILE_DELETE,
 } from '../constants/userManagement';
 import { ENDPOINTS } from '../utils/URL';
-import { UserStatus } from '../utils/enums';
+import { UserFinalDayStatus, UserStatus } from '../utils/enums';
```
Comment (EvianTan) [inline src/actions/userManagement.js:10]: `UserFinalDayStatus` is not existing, pls remove it.

File: src/actions/userManagement.js [hunk 2]
```diff
@@ -126,3 +132,29 @@ export const userProfileDeleteAction = (user) => {
     user,
   };
 };
+/**
+ * update the user profile
+ * @param {*} user - the user to be updated
+ * @param {*} finalDate  - the date to be inactive
+ */
+export const updateUserFinalDayStatus = (user,status,finalDayDate) => {
```
Comment (EvianTan) [inline src/actions/userManagement.js:140]: Pls add space trailing the commas, also add spaces around line 145 parentheses and around line 148 `else`, according to [Google JS style guide](https://google.github.io/styleguide/jsguide.html#formatting-braces)

File: src/components/UserManagement/SetUpFinalDayButton.jsx [hunk 3]
```diff
@@ -0,0 +1,74 @@
+import React, { useEffect, useState } from 'react';
+import { useDispatch } from 'react-redux';
+import { Button } from 'reactstrap';
+import { SET_FINAL_DAY, CANCEL } from '../../languages/en/ui';
+import SetUpFinalDayPopUp from './SetUpFinalDayPopUp';
+import { updateUserFinalDayStatus } from 'actions/userManagement';
+import axios from 'axios';
```
Comment (EvianTan) [inline src/components/UserManagement/SetUpFinalDayButton.jsx:7]: line 7 & 8 are not used, remove them.

File: src/components/UserManagement/SetUpFinalDayButton.jsx [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+import { ENDPOINTS } from 'utils/URL';
+import { toast } from 'react-toastify';
+/**
```
Comment (EvianTan) [inline src/components/UserManagement/SetUpFinalDayButton.jsx:10]: consider adding a break line before line 10

File: src/components/UserManagement/SetUpFinalDayButton.jsx [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+ * @returns
+ */
+const SetUpFinalDayButton = (props) => {
+  const[isSet, setIsSet] = useState(false);  
+  const[finalDayDateOpen, setFinalDayDateOpen] = useState(false);
+  const dispatch = useDispatch();
+
+  useEffect(() => {
+    if (props.userProfile?.endDate !== undefined)
+      setIsSet(true);
```
Comment (EvianTan) [inline src/components/UserManagement/SetUpFinalDayButton.jsx:23]: Add braces around line 23, according to https://google.github.io/styleguide/jsguide.html#formatting-braces

File: src/components/UserManagement/SetUpFinalDayButton.jsx [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+  }, [setIsSet]);
+  
+  const onFinalDayClick = (user, status) => {
+    if(isSet){
+      updateUserFinalDayStatus(props.userProfile,"Active",undefined)(dispatch)
+      setIsSet(!isSet);
+      toast.success('Your Final Day has been set.');
```
Comment (EvianTan) [inline src/components/UserManagement/SetUpFinalDayButton.jsx:30]: Consider using `This user's final day has been set`, similar apply to line 46

File: src/components/UserManagement/SetUpFinalDayButton.jsx [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    updateUserFinalDayStatus(props.userProfile, "Active", finalDayDate)(dispatch);
+    setIsSet(true);
+    setFinalDayDateOpen(false);
+    toast.success('Your Final Day has been deleted.');
+    window.location.reload();
+  }
+
+  return (
+    <React.Fragment>
+       <SetUpFinalDayPopUp
```
Comment (EvianTan) [inline src/components/UserManagement/SetUpFinalDayButton.jsx:52]: line 52 need rm one indent to match line 56

File: src/components/UserManagement/SetUpFinalDayPopUp.jsx [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+  const [finalDayDate, onDateChange] = useState(Date.now());
+  const [dateError, setDateError] = useState(false);
+
+  const closePopup = (e) => {
+    props.onClose();
+  };
+  const deactiveUser = () => {
+    if (Date.parse(finalDayDate) > Date.now()) {
+      props.onSave(finalDayDate);
+      //toast.success('Your Final Day has been set.');
```
Comment (EvianTan) [inline src/components/UserManagement/SetUpFinalDayPopUp.jsx:20]: remove this comment

File: src/components/UserProfile/BasicInformationTab/BasicInformationTab.jsx [hunk 9]
```diff
@@ -5,13 +5,13 @@ import moment from 'moment';
...
 import 'react-phone-input-2/lib/style.css';
-
 import PauseAndResumeButton from 'components/UserManagement/PauseAndResumeButton';
-
 import TimeZoneDropDown from '../TimeZoneDropDown';
 import { useSelector } from 'react-redux';
 import { getUserTimeZone } from 'services/timezoneApiService';
 import hasPermission from 'utils/permissions';
+import SetUpFinalDayButton from 'components/UserManagement/SetUpFinalDayButton';
+import ActiveCell from 'components/UserManagement/ActiveCell';
```
Comment (EvianTan) [inline src/components/UserProfile/BasicInformationTab/BasicInformationTab.jsx:14]: rm this line

**Comment by EvianTan** [comment]
@kseniaxu0704  You're doing awesome job on functionally changes!! All the funcs (set final day/ delete final day) are verified!
But there are two things you might need pay attention:
1. About JS code guide, we collaborate here, keep the same coding style could make other's life easier 😄 
2. That one might be a bug. When I set the final day to be 08/31/2022, it also shows 08/30/2022 on page, see below images. I'm guessing it might be a timezone problem. If it's need much effort to fix it, feel free to document it on our bug list.
image
image

File: src/actions/userManagement.js (hunk 1 above)
Comment (kseniaxu0704) [reply src/actions/userManagement.js:10]: thanks, i removed it

File: src/actions/userManagement.js [hunk 10]
```diff
@@ -44,15 +46,19 @@ export const updateUserStatus = (user, status, reactivationDate) => {
     patchData.endDate = undefined;
     userProfile.endDate = undefined;
   }
-
-  const updateProfilePromise = axios.patch(ENDPOINTS.USER_PROFILE(user._id), patchData);
+  const updateProfilePromise = axios.patch(ENDPOINTS.USER_PROFILE(user._id), patchData)
+  .then(function (response) {
```
[the remark this answers is not in the document]
Comment (kseniaxu0704) [reply src/actions/userManagement.js:50]: sure, change to the arrow fuctions

File: src/actions/userManagement.js (hunk 2 above)
Comment (kseniaxu0704) [reply src/actions/userManagement.js:140]: Thx, already added space after commas, around parentheses, and reserved keywords.

File: src/components/UserManagement/SetUpFinalDayButton.jsx (hunk 4 above)
Comment (kseniaxu0704) [reply src/components/UserManagement/SetUpFinalDayButton.jsx:10]: thx, add break line now on 10 & 74.

File: src/components/UserManagement/SetUpFinalDayButton.jsx (hunk 5 above)
Comment (kseniaxu0704) [reply src/components/UserManagement/SetUpFinalDayButton.jsx:23]: thx, move it in one line now

File: src/components/UserManagement/SetUpFinalDayButton.jsx (hunk 6 above)
Comment (kseniaxu0704) [reply src/components/UserManagement/SetUpFinalDayButton.jsx:30]: thx, already change form 'Your Final Day has been set.' to 'This user's final day has been set', also in line 46

File: src/components/UserManagement/SetUpFinalDayButton.jsx (hunk 7 above)
Comment (kseniaxu0704) [reply src/components/UserManagement/SetUpFinalDayButton.jsx:52]: thx for the comment, i will watch out the coding style more carefully.

File: src/components/UserManagement/SetUpFinalDayPopUp.jsx (hunk 8 above)
Comment (kseniaxu0704) [reply src/components/UserManagement/SetUpFinalDayPopUp.jsx:20]: thx, removed it

File: src/components/UserProfile/BasicInformationTab/BasicInformationTab.jsx (hunk 9 above)
Comment (kseniaxu0704) [reply src/components/UserProfile/BasicInformationTab/BasicInformationTab.jsx:14]: thx, i removed it

File: src/components/UserManagement/SetUpFinalDayButton.jsx (hunk 3 above)
Comment (kseniaxu0704) [reply src/components/UserManagement/SetUpFinalDayButton.jsx:7]: thx, already removed.

**Review by EvianTan — APPROVED** [review]

**Pull request merged**